cmd/cloud-controller-manager/nodeipamcontroller.go

```diff
@@ -180,50 +155,79 @@ func processCIDRs(cidrsList string) ([]*net.IPNet, bool, error) {
-	// NodeCIDRMaskSize can be used only for single stack clusters
-	if cfg.NodeCIDRMaskSize != 0 {
-		return ipv4Mask, ipv6Mask, errors.New("usage of --node-cidr-mask-size is not allowed with dual-stack clusters")
+func setNodeCIDRMaskSizes(cfg nodeipamconfig.NodeIPAMControllerConfiguration, clusterCIDRs []*net.IPNet) ([]int, error) {
+
+	sortedSizes := func(maskSizeIPv4, maskSizeIPv6 int) []int {
+		nodeMaskCIDRs := make([]int, len(clusterCIDRs))
+
+		for idx, clusterCIDR := range clusterCIDRs {
+			if netutils.IsIPv6CIDR(clusterCIDR) {
+				nodeMaskCIDRs[idx] = maskSizeIPv6
+			} else {
+				nodeMaskCIDRs[idx] = maskSizeIPv4
+			}
+		}
+		return nodeMaskCIDRs
 	}
-	if cfg.NodeCIDRMaskSizeIPv4 != 0 {
-		ipv4Mask = int(cfg.NodeCIDRMaskSizeIPv4)
+
+	// --node-cidr-mask-size flag is incompatible with dual stack clusters.
+	ipv4Mask, ipv6Mask := defaultNodeMaskCIDRIPv4, defaultNodeMaskCIDRIPv6
+	isDualstack := len(clusterCIDRs) > 1
+
+	// case one: cluster is dualstack (i.e, more than one cidr)
+	if isDualstack {
+		// if --node-cidr-mask-size then fail, user must configure the correct dual-stack mask sizes (or use default)
+		if cfg.NodeCIDRMaskSize != 0 {
+			return nil, errors.New("usage of --node-cidr-mask-size is not allowed with dual-stack clusters")
+		}
+
+		if cfg.NodeCIDRMaskSizeIPv4 != 0 {
+			ipv4Mask = int(cfg.NodeCIDRMaskSizeIPv4)
+		}
+		if cfg.NodeCIDRMaskSizeIPv6 != 0 {
+			ipv6Mask = int(cfg.NodeCIDRMaskSizeIPv6)
+		}
+		return sortedSizes(ipv4Mask, ipv6Mask), nil
 	}
-	if cfg.NodeCIDRMaskSizeIPv6 != 0 {
-		ipv6Mask = int(cfg.NodeCIDRMaskSizeIPv6)
+
+	maskConfigured := cfg.NodeCIDRMaskSize != 0
+	maskV4Configured := cfg.NodeCIDRMaskSizeIPv4 != 0
+	maskV6Configured := cfg.NodeCIDRMaskSizeIPv6 != 0
+	isSingleStackIPv6 := netutils.IsIPv6CIDR(clusterCIDRs[0])
+
+	// original flag is set
+	if maskConfigured {
+		// original mask flag is still the main reference.
+		if maskV4Configured || maskV6Configured {
+			return nil, errors.New("usage of --node-cidr-mask-size-ipv4 and --node-cidr-mask-size-ipv6 is not allowed if --node-cidr-mask-size is set. For dual-stack clusters please unset it and use IPFamily specific flags")
+		}
+
+		mask := int(cfg.NodeCIDRMaskSize)
+		return sortedSizes(mask, mask), nil
 	}
-	return ipv4Mask, ipv6Mask, nil
-}
 
-// getNodeCIDRMaskSizes is a helper function that helps the generate the node cidr mask
-// sizes slice based on the cluster cidr slice
-func getNodeCIDRMaskSizes(clusterCIDRs []*net.IPNet, maskSizeIPv4, maskSizeIPv6 int) []int {
-	nodeMaskCIDRs := make([]int, len(clusterCIDRs))
+	if maskV4Configured {
+		if isSingleStackIPv6 {
+			return nil, errors.New("usage of --node-cidr-mask-size-ipv4 is not allowed for a single-stack IPv6 cluster")
+		}
 
-	for idx, clusterCIDR := range clusterCIDRs {
-		if netutils.IsIPv6CIDR(clusterCIDR) {
-			nodeMaskCIDRs[idx] = maskSizeIPv6
-		} else {
-			nodeMaskCIDRs[idx] = maskSizeIPv4
+		if maskV6Configured {
```

Comment:
to get here I know that:
a) We are single-stack
b) `maskV4Configured` is true
c) `isSingleStackIPv6` is false

IOW v4 flag is set and this is a v4 cluster.  So this is checking that user didn't ALSO set v6 flag.

But L225 below checks that and since we know `isSingleStackIPv6` is false, the error at L227 will fire.

Ergo, you can remove L217-220 here as redundant